File: Data_Structure_through_C/Queue and Stack/queue_using_stack.c
/*
Author: Kumar Shanu
Program: Implement queue datastructure using stack.
*/

#include <stdio.h>
#include <stdlib.h>

#define N 10      // size of queue
int s1[N], s2[N]; // two stacks
int top1 = -1, top2 = -1;
int count = 0;

// push operation of stack
void push(int *stack, int x)
{
    if (stack == s1)
    {
        if (top1 == N - 1)
        {
            printf("Overflow");
            return;
        }
        top1++;
        s1[top1] = x;
    }
    if (stack == s2)
    {
        if (top2 == N - 1)
        {
            printf("Overflow");
            return;
        }
        top2++;
        s2[top2] = x;
    }
}

// pop operation of stack
int pop(int *stack)
{
    if (stack == s1)
    {
        if (top1 == -1)
        {
            printf("Underflow....");
        }
        else
            return s1[top1--];
    }
    if (stack == s2)
    {
        if (top2 == -1)
        {
            printf("Underflow....");
        }
        else
            return s2[top2--];
    }
}

// enqueue operation of queue
void enqueue(int x)
{
    if (top1 == N - 1)
    {
        printf("Queue is full");
        return;
    }
    // insert data into stack 1
    push(s1, x);
}

// dequeue operation of queue
int dequeue()
{
    int val, tmp;
    if (top1 == -1)
    {
        printf("Queue is Empty\n");
    }
    else
    {
        // remove elements from stack 1 and insert into stack 2
        while (top1 != -1)
        {
            tmp = pop(s1);
            push(s2, tmp);
        }

        // store value to return
        val = pop(s2);

        // move data back into stack 1
        while (top2 != -1)
        {
            tmp = pop(s2);
            push(s1, tmp);
        }
        return val;
    }
}

// print the queue
void display()
{
    printf("Queue is: \n");
    for (int i = 0; i <= top1; i++)
    {
        printf("%5d", s1[i]);
    }
    printf("\n");
}

// main function
int main()
{
    int choice, val;
    while (1)
    {
        printf("\nProgram of Queue using stack\n");
        printf("1. Enqueue\n2. Dequeue\n3. Display queue\n4. Exit");
        printf("\nEnter your choice: ");
        scanf("%d", &choice);
        switch (choice)
        {
        case 1:
            printf("Enter the value to insert: ");
            scanf("%d", &val);
            enqueue(val);
            break;
        case 2:
            val = dequeue();
            printf("%d", val);
            break;
        case 3:
            display();
            break;
        case 4:
            exit(0);
        }
    }
    return 0;
}

File: DataStructure.py/queue_using_stack.py
from Stack import Stack


class Queue(object):
    """Queue implementation using stcaks

    Args:
        size (int): size of the queue
    """

    def __init__(self, size=10):
        self.s1 = Stack(size)
        self.s2 = Stack(size)

    def is_empty(self):
        return self.s1.is_empty()

    def is_full(self):
        return self.s1.is_full()

    def enqueue(self, val):
        """Enqueue operation of queue

        Args:
            val (any): element to insert
        """
        if self.is_full():
            print("Queue is full")
            return
        self.s1.push(val)

    def dequeue(self):
        """
        Dequque operation of queue
        """
        if self.is_empty():
            print("Queue is Empty")
            return

        # move elements from stack 1 to stack 2
        while(not self.s1.is_empty()):
            val = self.s1.pop()
            self.s2.push(val)

        # store the value to return
        val = self.s2.pop()

        # move elements back to stack 1
        while (not self.s2.is_empty()):
            self.s1.push(self.s2.pop())
        return val

    def display(self):
        """
        Display/print the queue
        """
        print(self.s1.stk)


if __name__ == "__main__":
    q = Queue()
    while True:
        print("Queue Implementation using stack")
        print("1. Enqueue\n2. Dequeue\n3. Display\n4. Exit")
        choice = int(input("Enter your choice: "))
        if choice == 1:
            val = input("Enter value to insert: ")
            q.enqueue(val)
        elif choice == 2:
            print(q.dequeue())
        elif choice == 3:
            print("queue is: ")
            q.display()
        else:
            break
